h2o-core/src/test/java/water/runner/H2ORunner.java

```diff
@@ -0,0 +1,165 @@
+        eachNotifier.fireTestStarted();
+        try {
+            statement.evaluate();
+        } catch (AssumptionViolatedException e) {
+            eachNotifier.addFailedAssumption(e);
+        } catch (Throwable e) {
+            eachNotifier.addFailure(e);
+        } finally {
+            try {
+                checkLeakedKeys(description);
+            } catch (Throwable t) {
+                eachNotifier.addFailure(t);
+            }
+            eachNotifier.fireTestFinished();
+        }
+    }
+
+    private void checkLeakedKeys(Description description) {
+        CheckKeysTask checkKeysTask = new CheckKeysTask().doAllNodes();
+        if (checkKeysTask.leakedKeys.length == 0) {
+            return;
+        }
+
+        printLeakedKeys(checkKeysTask.leakedKeys);
+        throw new IllegalStateException(String.format("Test method '%s.%s' leaked %d keys.", description.getTestClass().getName(), description.getMethodName(), checkKeysTask.leakedKeys.length));
+    }
+
+
+    private void printLeakedKeys(final Key[] leakedKeys) {
+        Set<Key> leakedKeysSet = new HashSet<>(leakedKeys.length);
+
+        for (Key k : leakedKeys) {
+            leakedKeysSet.add(k);
+        }
+
+        for (Key key : leakedKeys) {
+
+            final Value keyValue = Value.STORE_get(key);
+            if (keyValue != null && keyValue.isFrame()) {
+                Frame frame = (Frame) key.get();
+                Log.err(String.format("Leaked frame with key '%s'. This frame contains the following vectors:", frame._key.toString()));
+
+                for (Key vecKey : frame.keys()) {
+                    if (!leakedKeysSet.contains(vecKey)) continue;
+                    Log.err(String.format("   Vector '%s'. This vector contains the following chunks:", vecKey.toString()));
+
+                    final Vec vec = (Vec) vecKey.get();
+                    for (int i = 0; i < vec.nChunks(); i++) {
+                        final Key chunkKey = vec.chunkKey(i);
+                        if (!leakedKeysSet.contains(chunkKey)) continue;
+                        Log.err(String.format("       Chunk id %d, key '%s'", i, chunkKey));
+                        leakedKeysSet.remove(chunkKey);
+                    }
+                    
+                    leakedKeysSet.remove(vecKey);
+                }
+                leakedKeysSet.remove(key);
+            }
+        }
+
+        if (!leakedKeysSet.isEmpty()) {
+            Log.err(String.format("%nThere are also %d more leaked keys:", leakedKeysSet.size()));
+        }
+
+        for (Key key : leakedKeysSet) {
+            Log.err(String.format("Key '%s'", key.toString()));
+        }
+    }
+
+
+
+    private int fetchCloudSize() {
+        final CloudSize annotation = testClass.getAnnotation(CloudSize.class);
+        if (annotation == null) throw new IllegalStateException("@CloudSize annotation is missing for test class: " + testClass.getName());
```

Comment:
why so restrictive, I think this should do nothing if the annotation is missing

Reply:
There was a discussion about it. The original design was `@Cloudsize(1)` behavior if there is no annotation. There were discussions and the main argument for this requirement was "lack of hidden magic".